File: MTG Deck Editor/src/editor/collection/export/DelimitedCardListFormat.java
package editor.collection.export;

import java.text.ParseException;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.StringJoiner;
import java.util.stream.Collectors;

import editor.collection.CardList;
import editor.collection.deck.Deck;
import editor.database.card.Card;
import editor.database.card.CardFormat;
import editor.database.characteristics.CardData;
import editor.gui.MainFrame;

/**
 * This class represents a formatter that creates a table whose columns
 * are card characteristics and whose rows contain the values of those
 * characteristics for the corresponding cards.
 * 
 * @author Alec Roelke
 */
public class DelimitedCardListFormat implements CardListFormat
{
	/**
	 * Default delimiter between table cells.  It is a comma.
	 */
	public static final String DEFAULT_DELIMITER = ",";
	/**
	 * Default data to present in the table.  By default, only a card's name,
	 * expansion, and copy count will be used.
	 */
	public static final List<CardData> DEFAULT_DATA = Arrays.asList(CardData.NAME,
			CardData.EXPANSION_NAME,
			CardData.COUNT);
	/**
	 * List of suggested delimiters.
	 */
	public static final String[] DELIMITERS = new String[] {",", ";", ":", "{tab}", "{space}"};
	/**
	 * String to surround cells that contain the delimiter with.
	 */
	public static final String ESCAPE = "\"";
	
	/**
	 * Split a string along a delimiter, but only if that delimiter is not inside {@value #ESCAPE}.
	 * If a split value is surrounded by {@value #ESCAPE}, remove them.
	 * 
	 * @param delimiter delimiter to split with
	 * @param line string to split
	 * @return an array containing the split string with delimiters and surrounding {@value #ESCAPE} removed
	 */
	public static String[] split(String delimiter, String line)
	{
		String[] cells = line.split(delimiter + "(?=(?:[^\"]*\"[^\"]*\")*[^\"]*$)");
		for (int i = 0; i < cells.length; i++)
			if (cells[i].substring(0, ESCAPE.length()).equals(ESCAPE) || cells[i].substring(cells[i].length() - ESCAPE.length()).equals(ESCAPE))
				cells[i] = cells[i].substring(1, cells[i].length() - 1);
		return cells;
	}
	
	/**
	 * Delimiter to separate table cells.
	 */
	private String delimiter;
	/**
	 * Whether or not to include column headers.  Parsing will require
	 * column headers.
	 */
	private boolean include;
	/**
	 * Data types to include in the table.
	 */
	private List<CardData> types;
	
	/**
	 * Create a new way to format a card list into a table.
	 * 
	 * @param delim delimiter for the table cells
	 * @param data data types to use for the columns
	 * @param headers whether or not to include column headers
	 */
	public DelimitedCardListFormat(String delim, List<CardData> data, boolean headers)
	{
		delimiter = delim;
		types = data;
		include = headers;
		
		if ("{space}".equals(delimiter))
			delimiter = " ";
		else if ("{tab}".equals(delimiter))
			delimiter = "\t";
	}
	
	/**
	 * Create a new DelimitedCardListFormat using the default values and
	 * including headers.
	 */
	public DelimitedCardListFormat()
	{
		this(DEFAULT_DELIMITER, DEFAULT_DATA, true);
	}
	
	@Override
	public String format(CardList list)
	{
		StringJoiner join = new StringJoiner(System.lineSeparator());
		List<CardFormat> columnFormats = types.stream().map((t) -> new CardFormat('{' + t.toString().toLowerCase() + '}')).collect(Collectors.toList());
		for (Card card: list)
		{
			StringJoiner line = new StringJoiner(delimiter);
			for (CardFormat format: columnFormats)
			{
				String value = format.format(list.getData(card));
				if (value.contains(delimiter))
					value = ESCAPE + value.replace(ESCAPE, ESCAPE + ESCAPE) + ESCAPE;
				line.add(value);
			}
			join.add(line.toString());
		}
		return join.toString();
	}
	
	@Override
	public String header()
	{
		if (include)
			return String.join(delimiter, types.stream().map(CardData::toString).toArray(String[]::new));
		else
			return "";
	}
	
	/**
	 * {@inheritDoc}
	 * @throws IllegalStateException if card name isn't a column
	 */
	@Override
	public CardList parse(String source) throws ParseException, IllegalStateException
	{
		List<String> lines = Arrays.stream(source.split(System.lineSeparator())).collect(Collectors.toList());
		
		int pos = 0;
		if (!include)
		{
			String[] headers = lines.get(0).split(delimiter);
			types = new ArrayList<CardData>(headers.length);
			for (String header: headers)
			{
				boolean success = false;
				for (CardData type: CardData.values())
				{
					if (header.compareToIgnoreCase(type.toString()) == 0)
					{
						types.add(type);
						success = true;
						break;
					}
				}
				if (!success)
					throw new ParseException("unknown data type " + header, pos);
				pos += header.length();
			}
			lines.remove(0);
		}
		
		int nameIndex = types.indexOf(CardData.NAME);
		if (nameIndex < 0)
			throw new IllegalStateException("can't parse cards without names");
		int expansionIndex = types.indexOf(CardData.EXPANSION_NAME);
		int numberIndex = types.indexOf(CardData.CARD_NUMBER);
		int countIndex = types.indexOf(CardData.COUNT);
		if (countIndex < 0)
			System.err.println("warning: missing card count in parse; assuming one copy of each card");
		int dateIndex = types.indexOf(CardData.DATE_ADDED);
		
		Deck deck = new Deck();
		for (String line: lines)
		{
			line = line.replace(ESCAPE + ESCAPE, ESCAPE);
			String[] cells = split(delimiter, line);
			
			List<Card> possibilities = MainFrame.inventory().stream().filter((c) -> c.unifiedName().equalsIgnoreCase(cells[nameIndex])).collect(Collectors.toList());
			if (possibilities.size() > 1 && expansionIndex > -1)
				possibilities.removeIf((c) -> !c.expansion().name.equalsIgnoreCase(cells[expansionIndex]));
			if (possibilities.size() > 1 && numberIndex > -1)
				possibilities.removeIf((c) -> !String.join(' ' + Card.FACE_SEPARATOR + ' ', c.number()).equals(cells[numberIndex]));
			
			if (possibilities.size() > 1)
				System.err.println("warning: cannot determine printing of " + possibilities.get(0).unifiedName());
			if (possibilities.isEmpty())
				throw new ParseException("can't find card named " + cells[nameIndex], pos);
			deck.add(possibilities.get(0), countIndex < 0 ? 1 : Integer.parseInt(cells[countIndex]), dateIndex < 0 ? LocalDate.now() : LocalDate.parse(cells[dateIndex], Deck.DATE_FORMATTER));
			pos += line.length();
		}
		
		return deck;
	}
	
}

File: src/main/scala/editor/collection/format/DelimitedCardListFormat.scala
package editor.collection.format

import com.mdimension.jchronic.Chronic
import editor.collection.CardList
import editor.collection.immutable.Inventory
import editor.collection.mutable.Deck
import editor.database.attributes.CardAttribute
import editor.database.card.Card
import editor.database.card.CardFormat
import editor.gui.MainFrame
import editor.gui.deck.DesignSerializer
import editor.unicode.{_, given}
import editor.util.IterableReader

import java.io.InputStream
import java.text.ParseException
import java.time.LocalDate
import java.time.ZoneId
import java.util.StringJoiner
import scala.collection.immutable.ListMap

/**
 * Object containing useful information and functions for formatting and parsing delimited files.
 * @author Alec Roelke
 */
object DelimitedCardListFormat {
  /** Default delimiter to use for separating table cells (a comma) */
  val DefaultDelimiter = ","

  /** Default list of attributes to use for columns (name, expansion, and count) */
  val DefaultData = Seq(CardAttribute.Name, CardAttribute.Expansion, CardAttribute.Count)

  /** Common delimiters used for separating cells on a line (comma, semicolon, colon, tab, and space) */
  val Delimiters = Seq(",", ";", ":", "{tab}", "{space}")

  /** String to use to escape delimiters (double quote) */
  val Escape = "\""

  /**
   * Split a string by a delimiter, ignoring delimiters contained inside [[Escape]] pairs.
   * 
   * @param delimiter delimiter to split by
   * @param line string to split
   * @return an array of strings containing the split line, with delimiters between tokens removed
   */
  def split(delimiter: String, line: String) = line.split(s"$delimiter(?=(?:[^$Escape]*$Escape[^$Escape]*$Escape)*[^$Escape]*$$)").collect{
    case cell if cell.substring(0, Escape.size) == Escape || cell.substring(cell.size - Escape.size) == Escape => cell.substring(1, cell.size - 1)
    case cell => cell
  }
}

/**
 * A card list formatter that formats cards into a table whose columns are [[CardAttribute]]s and whose rows
 * are the values of those attributes for individual [[Card]]s. It can also parse such a table into a card list,
 * making guesses about specific printings if the information provided doesn't narrow it down enough.
 * 
 * @constructor create a new card list formatter for a specific delimiter and list of attributes
 * @param delim delimiter to use to separate cells; to use a space or a tab, "{space}" and "{tab}" can be used
 * @param attributes list of attributes to include in the table in the order they should appear; use an empty list to elide headers from
 * the output or infer them from input when parsing
 * 
 * @author Alec Roelke
 */
class DelimitedCardListFormat(delim: String, attributes: Seq[CardAttribute[?, ?]]) extends CardListFormat {
  import DelimitedCardListFormat._

  private val delimiter = delim match {
    case "{space}" => " "
    case "{tab}" => "\t"
    case _ => delim
  }

  override def format(list: CardList) = {
    val columnFormats = attributes.map((a) => CardFormat(s"{$a}".toLowerCase))
    list.map((e) => {
      columnFormats.map((format) => {
        val value = format.format(e).replace(Escape.toString, Substitute)
        (if (value.contains(delimiter)) s"$Escape${value.replace(Escape, Escape*2)}$Escape" else value).replace(Substitute, Escape*2)
      }).mkString(delimiter)
    }).mkString(System.lineSeparator)
  }

  override lazy val header = Option.when(!attributes.isEmpty)(attributes.map(_.toString.replace(Escape, Escape*2)).mkString(delimiter))

  override def parse(source: InputStream) = {
    val deck = Deck()
    var extra: Option[String] = None
    var extras = ListMap[String, Deck]()
    var pos = 0

    var name = attributes.indexOf(CardAttribute.Name)
    var expansion = attributes.indexOf(CardAttribute.Expansion)
    var number = attributes.indexOf(CardAttribute.CardNumber)
    var count = attributes.indexOf(CardAttribute.Count)
    var date = attributes.indexOf(CardAttribute.DateAdded)
    IterableReader(source).foreach((line) => {
      if (name < 0) {
        val headers = line.split(delimiter)
        val attrs = headers.map((h) => CardAttribute.displayableValues.find(_.toString.equalsIgnoreCase(h)).getOrElse(throw ParseException(s"unknown data type $h", pos))).toSeq
        name = attrs.indexOf(CardAttribute.Name)
        expansion = attrs.indexOf(CardAttribute.Expansion)
        number = attrs.indexOf(CardAttribute.CardNumber)
        count = attrs.indexOf(CardAttribute.Count)
        date = attrs.indexOf(CardAttribute.DateAdded)
        if (name < 0)
          throw IllegalStateException("can't parse cards without names")
        if (count < 0)
          System.err.println("warning: missing card count in parse; assuming one copy of each card")
      } else {
        try {
          val cells = split(delimiter, line.replace(Escape*2, Escape))
          val possibilities = Inventory
              .filter(_.card.name.equalsIgnoreCase(cells(name)))
              .filter(expansion < 0 || _.card.expansion.name.equalsIgnoreCase(cells(expansion)))
              .filter(number < 0 || _.card.faces.map(_.number).mkString(Card.FaceSeparator) == cells(number))
          
          if (possibilities.size > 1)
            System.err.println(s"warning: cannot determine printing of \"${line.trim}\"")
          if (possibilities.isEmpty)
            throw ParseException(s"can't find card named ${cells(name)}", pos)
          
          extra.map(extras).getOrElse(deck).add(
            possibilities.head.card,
            if (count < 0) 1 else cells(count).toInt,
            if (date < 0) LocalDate.now else try {
              Option(Chronic.parse(cells(date)))
                  .map(_.getBeginCalendar.getTime.toInstant.atZone(ZoneId.systemDefault).toLocalDate)
                  .getOrElse(LocalDate.now)
            } catch case _: IllegalStateException => LocalDate.now
          )
        } catch case e: ParseException => {
          extra = Some(line)
          extras += line -> Deck()
        }
      }
      pos += line.size
    })

    DesignSerializer(deck, extras, "", "")
  }
}
